### PR TITLE
integrate change from src-d/ml nn tokenizer

### PR DESCRIPTION
**Pull request by glimow**
Adding the NN id splitter to src-d/ml-core. PR migrated from https://github.com/src-d/ml/pull/405
Note that I separated original PR into two. This one only adds the neural network model code, but do not modifiy the `tokenparser` so it can be merged before https://github.com/src-d/models/pull/18

### CODE REVIEW

File: sourced/ml/core/models/id_splitter.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+import logging
+import string
+from typing import List, Sequence, Tuple
+
+import keras
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: We should use the Tensorflow implementation instead of the original one. https://www.tensorflow.org/guide/keras

File: sourced/ml/core/models/id_splitter.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            for char, label in zip(clean_id, id_output):
+                if label == 1:
+                    splitted_ids.append(splitted_id)
+                    splitted_id = ""
+                splitted_id += char
+            splitted_ids.append(splitted_id)
+        return splitted_ids
+
+    def __del__(self):
+        self.clear_keras()
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: Tensorflow implementation should work fine without the need to clear sessions

File: .gitignore [hunk 3]
```diff
@@ -113,4 +113,7 @@ ENV/
 .mypy_cache/
 
 # CI
-.ci
\ No newline at end of file
+.ci
+
+#vscode
```
Comment (vmarkovtsev) [inline .gitignore]:
```suggestion
```

(kinda obvious)

File: sourced/ml/core/models/id_splitter.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+from modelforge import Model, register_model
+import numpy
+
+from sourced.ml.core.algorithms.id_splitter.nn_model import (f1score, precision,
+                                                             recall)
+from sourced.ml.core.models.license import DEFAULT_LICENSE
+
+
+@register_model
+class IdentifierSplitterNN(Model):
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
class IdentifierSplitterBiLSTM(Model):
```

File: sourced/ml/core/models/id_splitter.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    Bi-Directionnal LSTM Model. Splits identifiers without need for a conventional pattern.
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]:
```
Reference: https://arxiv.org/abs/1805.11651
```

File: sourced/ml/core/models/id_splitter.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    NAME = "id_splitter_nn"
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
    NAME = "id_splitter_bilstm"
```

File: sourced/ml/core/models/id_splitter.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    LICENSE = DEFAULT_LICENSE
+
+    DEFAULT_MAXLEN = 40
+    DEFAULT_PADDING = "post"
+    DEFAULT_MAPPING = None
+
+    def construct(self, model: "keras.models.Model",
+                  maxlen: int = DEFAULT_MAXLEN,
+                  padding: str = DEFAULT_PADDING,
+                  mapping: dict = DEFAULT_MAPPING) -> "IdentifierSplitterNN":
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: Would be great to have a docstring with argument descriptions

File: sourced/ml/core/models/id_splitter.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        return self
+
+    @property
+    def model(self) -> "keras.models.Model":
+        """
+        Returns the wrapped keras model.
+        """
+        return self._model
+
+    def clear_keras(self):
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: We should remove this and call `keras.backend.clear_session()` directly

File: sourced/ml/core/models/id_splitter.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        return {
+            "config": self._model.get_config(),
+            "weights": self._model.get_weights(),
+            "mapping": self._mapping,
+            }
+
+    def _load_tree(self, tree: dict):
+        model = keras.models.Model.from_config(tree["config"])
+        model.set_weights(tree["weights"])
+        if "mapping" in tree:
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: A mapping should always be present

File: sourced/ml/core/models/id_splitter.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+    def _prepare_single_identifier(self, identifier: str) -> Tuple[numpy.array, str]:
+        if self._mapping is None:
+            self._mapping = {c: i for i, c in enumerate(string.ascii_lowercase, start=1)}
+
+        # Clean identifier
+        clean_id = "".join(char for char in identifier.lower() if char in self._mapping)
+        if len(clean_id) > self._maxlen:
+            clean_id = clean_id[:self._maxlen]
+        logging.info("Preprocessed identifier: {}".format(clean_id))
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: Define a class-level `_log = logging.getLogger("IdentifierSplitterBiLSTM")` and use it for logging.

Comment (vmarkovtsev) [reply sourced/ml/core/models/id_splitter.py]: The log level should be debug

Comment (vmarkovtsev) [reply sourced/ml/core/models/id_splitter.py]: We use only `%` ("old") style string substitutions. https://github.com/src-d/guide/blob/master/engineering/conventions/python.md#code-style

Comment (vmarkovtsev) [reply sourced/ml/core/models/id_splitter.py]: Please include the original string into the log message

File: sourced/ml/core/models/id_splitter.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    DEFAULT_MAXLEN = 40
+    DEFAULT_PADDING = "post"
+    DEFAULT_MAPPING = None
+
+    def construct(self, model: "keras.models.Model",
+                  maxlen: int = DEFAULT_MAXLEN,
+                  padding: str = DEFAULT_PADDING,
+                  mapping: dict = DEFAULT_MAPPING) -> "IdentifierSplitterNN":
+
+        self._maxlen = maxlen
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: `maxlen` and `padding` can be inferred from the model. Make them properties.

File: sourced/ml/core/models/id_splitter.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    def _load_tree(self, tree: dict):
+        model = keras.models.Model.from_config(tree["config"])
+        model.set_weights(tree["weights"])
+        if "mapping" in tree:
+            self.construct(model, mapping=tree["mapping"])
+        else:
+            self.construct(model)
+
+    def _prepare_single_identifier(self, identifier: str) -> Tuple[numpy.array, str]:
+        if self._mapping is None:
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: We need to check that our mapping plays nicely with `int8` on line 77

Comment (vmarkovtsev) [reply sourced/ml/core/models/id_splitter.py]: `mapping` should be initialized at construction or load time

File: sourced/ml/core/models/id_splitter.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        self._model = keras.models.load_model(path, custom_objects={"precision": precision,
+                                                                    "recall": recall,
+                                                                    "f1score": f1score})
+
+    def split(self, identifiers: Sequence[str]) -> List[str]:
+        """
+        Splits a lists of identifiers using the model.
+        """
+        feats, clean_ids = self.prepare_input(identifiers)
+        output = self._model.predict(feats, batch_size=4096)
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]: Batch size should be configurable (class level constant). Imagine that 4096 does not fit the GPU memory - then our model will crash and not work at all

File: .gitignore (hunk 3 above)
Comment (glimow) [reply .gitignore]: Sure, but I just followed the conventions I have seen for this file

Comment (vmarkovtsev) [reply .gitignore]: Some of the comments should be removed from the original file, that's true. It was auto-added in GitHub UI, and nobody looked deep inside. So:

- `#MacOS` -> `# macOS`
- `#PyCharm IDE` -> `# PyCharm IDE`
- `* stuff:` -> `*`
- Delete `# Scrapy stuff:`
- Delete `# dotenv`
- Delete `# mypy`
- Delete `# CI`

File: sourced/ml/core/models/id_splitter.py (hunk 13 above)
Comment (glimow) [reply sourced/ml/core/models/id_splitter.py]: nice catch

File: sourced/ml/core/models/id_splitter.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    DEFAULT_MAXLEN = 40
+    DEFAULT_PADDING = "post"
+    DEFAULT_MAPPING = {c: i for i, c in enumerate(string.ascii_lowercase, start=1)}
+    DEFAULT_BATCH_SIZE = 4096
+
+    def construct(self, model: "keras.models.Model",
+                  maxlen: int = DEFAULT_MAXLEN,
+                  padding: str = DEFAULT_PADDING,
+                  mapping: dict = DEFAULT_MAPPING,
+                  batch_size: int = DEFAULT_BATCH_SIZE) -> "IdentifierSplitterNN":
```
Comment (zurk) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
                  batch_size: int = DEFAULT_BATCH_SIZE) -> "IdentifierSplitterBiLSTM":
```

File: sourced/ml/core/models/id_splitter.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+    DEFAULT_MAXLEN = 40
+    DEFAULT_PADDING = "post"
+    DEFAULT_MAPPING = {c: i for i, c in enumerate(string.ascii_lowercase, start=1)}
+    DEFAULT_BATCH_SIZE = 4096
+
+    def construct(self, model: "keras.models.Model",
+                  maxlen: int = DEFAULT_MAXLEN,
+                  padding: str = DEFAULT_PADDING,
+                  mapping: dict = DEFAULT_MAPPING,
```
Comment (zurk) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
                  mapping: Dict[str, int] = DEFAULT_MAPPING,
```

File: sourced/ml/core/models/id_splitter.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+                  batch_size: int = DEFAULT_BATCH_SIZE) -> "IdentifierSplitterNN":
+        """
+        :param model: keras model used for identifier splitting.
+        :param maxlen: maximum length of input identifers.
+        :param padding: where to pad the identifiers of length < maxlen. Can be "left" or "right".
+        :param mapping: { str: int } mapping of characters to integers.
```
Comment (zurk) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
        :param mapping: mapping of characters to integers.
```

File: sourced/ml/core/models/id_splitter.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    VENDOR = "source{d}"
+    DESCRIPTION = "Weights of the BiLSTM network to split source code identifiers."
+    LICENSE = DEFAULT_LICENSE
+
+    DEFAULT_MAXLEN = 40
+    DEFAULT_PADDING = "post"
+    DEFAULT_MAPPING = {c: i for i, c in enumerate(string.ascii_lowercase, start=1)}
+    DEFAULT_BATCH_SIZE = 4096
+
+    def construct(self, model: "keras.models.Model",
```
Comment (zurk) [inline sourced/ml/core/models/id_splitter.py]: @vmarkovtsev in the last modelforge release I do not see `construct` as an abstract method. I think it is kind of deprecated approach and this code should be in`__init__` of the model as we have for https://github.com/src-d/style-analyzer/blob/1da666ab60a12da2c6aab27f736ce6be9354d1b7/lookout/style/typos/model.py#L16-L19 
WDYT?

File: sourced/ml/core/models/id_splitter.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+            }
+
+    def _load_tree(self, tree: dict):
+        model = keras.models.Model.from_config(tree["config"])
+        model.set_weights(tree["weights"])
+        self.construct(model, maxlen=tree["maxlen"],
+                       padding=tree["padding"], mapping=tree["mapping"])
+
+    def _prepare_single_identifier(self, identifier: str) -> Tuple[numpy.array, str]:
+
```
Comment (zurk) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
```

Strange that it passed style check 🤔

File: sourced/ml/core/models/id_splitter.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+        # Clean identifier
+        clean_id = "".join(char for char in identifier.lower() if char in self._mapping)
+        if len(clean_id) > self._maxlen:
+            clean_id = clean_id[:self._maxlen]
+        self._log.debug("Preprocessed identifier: %s : %s" % (identifier, clean_id))
+        return numpy.array([self._mapping[c] for c in clean_id]), clean_id
+
+    def prepare_input(self, identifiers: Sequence[str]) -> Tuple[numpy.array, List[str]]:
+        """
+        Prepares input by converting a sequence of identifiers to the corresponding
```
Comment (zurk) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
        Prepare input by converting a sequence of identifiers to the corresponding \
```

Again. Something is definitely wrong with the check.

File: sourced/ml/core/models/id_splitter.py (hunk 18 above)
Comment (zurk) [reply sourced/ml/core/models/id_splitter.py]: ok, it looks like documentation checks are not integrated yet.

File: sourced/ml/core/models/id_splitter.py (hunk 17 above)
Comment (vmarkovtsev) [reply sourced/ml/core/models/id_splitter.py]: `__init__` declares the empty attributes while `construct` applies the set of new ones so that the object becomes consistent again. Here it is fine.

File: sourced/ml/core/models/id_splitter.py (hunk 19 above)
Comment (vmarkovtsev) [reply sourced/ml/core/models/id_splitter.py]: Yeah that should have been checked.

File: sourced/ml/core/models/id_splitter.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+from sourced.ml.core.algorithms.id_splitter.nn_model import (f1score, precision,
+                                                             recall)
+from sourced.ml.core.models.license import DEFAULT_LICENSE
+
+
+@register_model
+class IdentifierSplitterBiLSTM(Model):
+    """
+    Bi-Directionnal LSTM Model. Splits identifiers without need for a conventional pattern.
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
    Bidirectional LSTM Model. Splits identifiers without need for a conventional pattern.
```

File: sourced/ml/core/models/id_splitter.py (hunk 17 above)
Comment (zurk) [reply sourced/ml/core/models/id_splitter.py]: ok, then we should add `__init__ ` with empty attributes declaration.

File: sourced/ml/core/models/id_splitter.py (hunk 19 above)
Comment (zurk) [reply sourced/ml/core/models/id_splitter.py]: I find out why. We have a lot of style mistakes from an old and not so good style check. 
So a lot of modules are just ignored. 
https://github.com/src-d/ml-core/blob/master/.flake8

It is a lot of work to fix all style issues...

**Comment by vmarkovtsev** [comment]
@glimow Commit messages: https://github.com/src-d/ml-core/blob/master/docs/CONTRIBUTING.md#format-of-the-commit-message

File: sourced/ml/core/models/id_splitter.py (hunk 17 above)
Comment (glimow) [reply sourced/ml/core/models/id_splitter.py]: there is no __init__ method used in any of the current modelforge models, however this method is defined in the base class `Model`. Shall I overload it ? Do we have to upgrade every different model ?

**Comment by glimow** [comment]
@vmarkovtsev I thought I started following the guidelines about the commit messages already. Do you want me to edit all old commit messages to add the cap at the first letter ?

File: sourced/ml/core/models/id_splitter.py (hunk 17 above)
Comment (vmarkovtsev) [reply sourced/ml/core/models/id_splitter.py]: You need to call `super()`, yes

**Comment by vmarkovtsev** [comment]
Yes, run an interactive rebase and reword all the commits.

File: sourced/ml/core/models/id_splitter.py [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+from sourced.ml.core.algorithms.id_splitter.nn_model import (f1score, precision,
+                                                             recall)
+from sourced.ml.core.models.license import DEFAULT_LICENSE
+
+
+@register_model
+class IdentifierSplitterBiLSTM(Model):
+    """
+    Bidirectionnal LSTM Model. Splits identifiers without need for a conventional pattern.
```
Comment (vmarkovtsev) [inline sourced/ml/core/models/id_splitter.py]:
```suggestion
    Bidirectional LSTM Model. Splits identifiers without need for a conventional pattern.
```
I wonder if I missed this last time or it was added afterward...

Comment (glimow) [reply sourced/ml/core/models/id_splitter.py]: Probably my bad. Fixed.

**Comment by zurk** [comment]
@glimow can I ask you to rebase your commits to the current master and I will merge?

**Comment by glimow** [comment]
Sure. It should be done.
